### DISCUSSION TITLE
SDL Unable to call Window.GetView

### DISCUSSION THREAD

**Discussion by TheVave** [dotnet/Silk.NET#1963] [Q&A]
# Summary
I have been trying to use SDL to create a window (or in this case IView) in Android, but it is giving me the following error:
`Silk.NET.SDL.SdlException: 'Application didn't initialize properly, did you include SDL_main.h in the file containing your main() function?'`
This looks like an error from SDL, so I'm not sure what the cause would be.
Here's my code (that probably has something wrong):
```
var options = ViewOptions.Default;
options.API = new GraphicsAPI(ContextAPI.OpenGLES, ContextProfile.Core, ContextFlags.Default, new APIVersion(3, 2));
vw = Window.GetView(options);
vw.Run();
````
vw is an IView object.

# Steps to reproduce (With above code)
- Platform: Android Mobile Emulator Pixel 5
- Framework Version: .NET 8 & MAUI
- API: OpenGLES 3.2

**Comment by Perksey** [comment]
What’s the code for your activity?

**Reply by TheVave** [reply]
The activity just calls the code above.

**Reply by Perksey** [reply]
In what function? Can you send the code please.

**Reply by TheVave** [reply]
Sorry for the delay.
MainPage.xml:
```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="PhoneTester.MainPage">

    <ScrollView>
        <VerticalStackLayout>
            <VerticalStackLayout VerticalOptions="Center" HorizontalOptions="Center" IsVisible="false" x:Name="Loading">
                <ActivityIndicator IsVisible="true" WidthRequest="100" HeightRequest="100" IsRunning="True"/>
                <Label Text="Loading..." HorizontalOptions="Center"/>
            </VerticalStackLayout>
            <Button Text="Start" Clicked="Start_Clicked"/>
        </VerticalStackLayout>
    </ScrollView>

</ContentPage>
```
MainPage.xml.cs:
```
using SharpPhysics._2d.Rendering._2DSGLESRenderer.Windowing;

namespace PhoneTester
{
	public partial class MainPage : ContentPage
	{

		public MainPage()
		{
			InitializeComponent();
		}

		private void Start_Clicked(object sender, EventArgs e)
		{
			Loading.IsVisible = true;
			((Button)sender).IsVisible = true;
			SDLTester.Run();
		}
	}

}
```
SDLTester class:
```
using Silk.NET.Windowing.Sdl;
using Silk.NET.Windowing;
using SharpPhysics._2d._2DSGLRenderer;
namespace SharpPhysics._2d.Rendering._2DSGLESRenderer.Windowing
{
	public class SDLTester
	{
		public static Size WndSze { get; internal set; }
		public static string WndTitle { get; internal set; }
		public static IView vw { get; internal set; }

		public static unsafe void Run()
		{
			var options = ViewOptions.Default;
			options.API = new GraphicsAPI(ContextAPI.OpenGLES, ContextProfile.Core, ContextFlags.Default, new APIVersion(3, 2));
			vw = Window.GetView(options);
			vw.Run();
		}
	}
}
```
I'm not calling any Silk.NET methods in my code other than inside the SDLTester.Run().

**Reply by Perksey** [reply]
We don’t support MAUI at this time. If you’d like to target Android, you have to use native .NET 8 Android (without MAUI) as demonstrated in the demo in the examples folder.

**Reply by TheVave** [reply]
I'm trying to use Maui as just the loading ui, and create a separate window for the actual program, and the only way I was aware of to get the android target was through Xamarin, which going to MAUI, and MAUI itself. 
Basically, I just want to use MAUI for the build target, and not for the window, where I'm using SDL.

**Reply by Perksey** [reply]
The .NET marketing people aren’t great. Not all .NET on Mobile is MAUI, MAUI refers to the XAML-based UI framework that has been ported to .NET formerly known as Xamarin Forms. Xamarin Android (MonoDroid) and Xamarin iOS (MonoTouch) i.e. the native counterparts are just known as .NET Android and .NET iOS respectively. We only support .NET Android.

The Android demo in the examples directory in the repo shows an example of using .NET Android with Silk.NET.

**Reply by TheVave** [reply]
So, should I create an "Android Application" in the VS project creator?

**Reply by Perksey** [reply]
Yes, assuming that makes a modern .NET project and not a Xamarin one.

**Reply by Perksey** [reply]
If the resultant csproj contains `net8.0-android` then it has created the correct type of project.

**Reply by TheVave** [reply]
Well, making a Android app and calling SDLTester.Run and all of the default things for a Android app, this is my activity:
```
namespace PhoneTester
{
	[Activity(Label = "@string/app_name", MainLauncher = true)]
	public class MainActivity : Activity
	{
		protected override void OnCreate(Bundle? savedInstanceState)
		{
			base.OnCreate(savedInstanceState);

			// Set our view from the "main" layout resource
			SetContentView(Resource.Layout.activity_main);

			// load new SDL window
			SDLTester.Run();
		}
	}
}
```
With the same code for SDL, that still doesn't work:
```
namespace SharpPhysics._2d.Rendering._2DSGLESRenderer.Windowing
{
	public class SDLTester
	{
		public static Size WndSze { get; internal set; }
		public static string WndTitle { get; internal set; }
		public static IView vw { get; internal set; }

		public static unsafe void Run()
		{
			var options = ViewOptions.Default;
			options.API = new GraphicsAPI(ContextAPI.OpenGLES, ContextProfile.Core, ContextFlags.Default, new APIVersion(3, 2));
			vw = Silk.NET.Windowing.Window.GetView(options);
			vw.Run();
		}
	}
}
```

**Reply by Perksey** [reply]
Have you looked at the demo in the examples folder of the repo that I suggested?

**Reply by TheVave** [reply]
Looking [here](https://github.com/dotnet/Silk.NET/blob/main/examples/CSharp/OpenGL%20Demos/AndroidDemo/MainActivity.cs), my code just crashes at line 61 essentially.

**Reply by Perksey** [reply]
Weird, and you're calling it on your overridden OnRun function as well?

**Reply by TheVave** [reply]
I wasn't using the actual demos, I was still calling my own code, but the demos have the exact same logic for windowing, so I am unsure of why it's happening. I think I am also going to try to put it on an actual phone, see if that helps.

**Reply by Perksey** [reply]
That doesn't answer my question, but instead of slow-walking to why you're not following the demo I'll just get straight to it: you need to inherit from SilkActivity and then override the OnRun function. OnRun then essentially acts as your new main.

**Reply by TheVave** [reply]
MySilkActivity.cs:
```
public class MySilkActivity : SilkActivity
{
	protected override void OnRun()
	{
		IView vw;

		var options = ViewOptions.Default;
		options.API = new GraphicsAPI(ContextAPI.OpenGLES, ContextProfile.Compatability, ContextFlags.Default, new APIVersion(3, 0));
		vw = Silk.NET.Windowing.Window.GetView(options);
		vw.Run();
	}
}
```
I decided to remove the library just for simplicity, and I know this is a little bit of a silly question, but OnRun isn't being invoked, should it be called by me or C# in some way?

**Reply by Perksey** [reply]
It is called by SDL when your activity is opened. Is Android definitely opening this activity?

**Reply by TheVave** [reply]
It doesn't appear so, should I remove the MainActivity class?

**Reply by Perksey** [reply]
I recommend familiarising with how Android development works. https://learn.microsoft.com/en-us/dotnet/maui/android/manifest?view=net-maui-8.0

**Reply by Perksey** [reply]
Generally I just replace the MainActivity rather than removing it because I can't be bothered to deal with the manifest complications etc, but that link should explain more.

**Reply by TheVave** [reply]
Testing with the following code for MainActivity.cs:
```
using Silk.NET.Windowing.Sdl.Android;
using Silk.NET.Windowing;
namespace PhoneTester;

[Activity(Label = "Debugging Application", MainLauncher = true)]
public class MainActivity : SilkActivity
{
	protected override void OnRun()
	{
		IView vw;

		var options = ViewOptions.Default;
		options.API = new GraphicsAPI(ContextAPI.OpenGLES, ContextProfile.Compatability, ContextFlags.Default, new APIVersion(3, 0));
		vw = Silk.NET.Windowing.Window.GetView(options);
		vw.Run();
	}
}
```
The above throws an exception about needing Bluetooth permission, yet when I go to settings it has that permission.
Also, I added a required permission of Bluetooth to the Required Permissions box when opening AndroidManifest.xml, and it still didn't work correctly.
Is this normal?

**Reply by Perksey** [reply]
Make sure that you have BLUETOOTH_CONNECT in your manifest and that you're targeting API 31 or later. If the issue persists let me know, we have seen this before but we haven't had reports of it since we updated SDL. Make sure you're using the latest version of all silk packages also.

**Reply by TheVave** [reply]
This is my AndroidManifest.xml (in xml viewer):
```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android">
	<application android:allowBackup="true" android:icon="@mipmap/appicon" android:label="@string/app_name" android:supportsRtl="true"></application>
	<uses-permission android:name="android.permission.INTERNET" />
	<uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
	<uses-sdk />
</manifest>
```
and it still throws the following error:
```
Java.Lang.RuntimeException: 'Unable to start activity ComponentInfo{com.companyname.pntstr/crc64893e06940e61b983.MainActivity}: java.lang.SecurityException: Need android.permission.BLUETOOTH_CONNECT permission for AttributionSource { uid = 10186, packageName = com.companyname.pntstr, attributionTag = null, token = android.os.BinderProxy@f384ee7, next = null }: AdapterService getBondedDevices'
```
I checked the packages and NuGet said they were up to date at 2.20.0.

PS: com.companyname.pntstr is the package name for my app package
